lib/spack/spack/fetch_strategy.py

```diff
@@ -55,74 +55,113 @@
 import spack.error
 import spack.util.crypto as crypto
 import spack.util.pattern as pattern
+
+from spack.util.compression import decompressor_for, extension
 from spack.util.executable import which
 from spack.util.string import comma_or
 from spack.version import Version, ver
-from spack.util.compression import decompressor_for, extension
 
-
-#: List of all fetch strategies, created by FetchStrategy metaclass.
+#: List of all known fetch strategies
 all_strategies = []
 
 
-def _needs_stage(fun):
-    """Many methods on fetch strategies require a stage to be set
-       using set_stage().  This decorator adds a check for self.stage."""
+def fetcher(cls):
+    """Registers a fetch strategy
+
+    Args:
+        cls (FetchStrategy): fetcher class to be registered
+    """
+    all_strategies.append(cls)
+    return cls
+
+
+def has_expanded_tree(directory):
+    """Checks if the input directory has at least one other
+    child subdirectory.
 
-    @wraps(fun)
-    def wrapper(self, *args, **kwargs):
-        if not self.stage:
-            raise NoStageError(fun)
-        return fun(self, *args, **kwargs)
+    Args:
+         directory (path): directory where to search
 
-    return wrapper
+    Returns:
+        first subdirectory found or None
+    """
+    for p in [os.path.join(directory, f) for f in os.listdir(directory)]:
+        if os.path.isdir(p):
+            return p
+    return None
 
 
-class FSMeta(type):
-    """This metaclass registers all fetch strategies in a list."""
-    def __init__(cls, name, bases, dict):
-        type.__init__(cls, name, bases, dict)
-        if cls.enabled:
-            all_strategies.append(cls)
+class FetchStrategy(object):
+    """Defines the interface that all fetchers need to comply with"""
 
+    #: Attributes required in version() args
+    required_attributes = tuple()
+    #: If set it will be used to save an archive to a specific filename
+    save_filename = None
+    #: Private cache for source_dir, initialized to None in the class
+    _source_dir = None
+    #: Private cache for expected_archive_files,
+    #: initialized to None in the class
+    _expected_archive_files = None
 
-class FetchStrategy(with_metaclass(FSMeta, object)):
-    """Superclass of all fetch strategies."""
-    enabled = False  # Non-abstract subclasses should be enabled.
-    required_attributes = None  # Attributes required in version() args.
+    @property
+    def source_dir(self):
```

Comment:
I think we should rename this.  I find `source_dir` really confusing because it's actually the root of the stage here, not the expanded archive's source directory.  Why not `stage_path`?

I'm wondering whether this isn't still too entangled.  I think that maybe `stage` should be responsible for all directory management, and the fetcher should just assume that it is fetching into the current working directory all the time... That might simplify a lot of the logic here...  I need to think about this more.

Reply:
Doesn't this mean that the client of a fetcher is responsible of moving to the correct directory, and then moving back when he's finished? I think it may simplify the logic here, not it's use within the code.